Libs/Optimize/ParticleSystem/VtkMeshWrapper.cpp

```diff
@@ -0,0 +1,822 @@
+  this->poly_data_->BuildCells();
+  this->poly_data_->BuildLinks();
+
+  vtkSmartPointer<vtkGenericCell> cell = vtkSmartPointer<vtkGenericCell>::New();
+  for (int i = 0; i < this->poly_data_->GetNumberOfCells(); i++) {
+    this->poly_data_->GetCell(i, cell);
+
+    vtkSmartPointer<vtkTriangle> triangle = vtkSmartPointer<vtkTriangle>::New();
+    triangle->GetPointIds()->SetNumberOfIds(3);
+    triangle->GetPointIds()->SetId(0, cell->GetPointId(0));
+    triangle->GetPointIds()->SetId(1, cell->GetPointId(1));
+    triangle->GetPointIds()->SetId(2, cell->GetPointId(2));
+    triangle->GetPoints()->SetPoint(0, cell->GetPoints()->GetPoint(0));
+    triangle->GetPoints()->SetPoint(1, cell->GetPoints()->GetPoint(1));
+    triangle->GetPoints()->SetPoint(2, cell->GetPoints()->GetPoint(2));
+
+    this->triangles_.push_back(triangle);
+  }
+
+  this->cell_locator_ = vtkSmartPointer<vtkCellLocator>::New();
+  this->cell_locator_->SetCacheCellBounds(true);
+  this->cell_locator_->SetDataSet(this->poly_data_);
+  this->cell_locator_->BuildLocator();
+
+  this->ComputeMeshBounds();
+  this->ComputeGradN();
+}
+
+//---------------------------------------------------------------------------
+double VtkMeshWrapper::ComputeDistance(VtkMeshWrapper::PointType pointa,
+                                       VtkMeshWrapper::PointType pointb) const
+{
+  double x = pointa[0] - pointb[0];
+  double y = pointa[1] - pointb[1];
+  double z = pointa[2] - pointb[2];
+  double distance = std::sqrt((x * x) + (y * y) + (z * z));
+  return distance;
+}
+
+//---------------------------------------------------------------------------
+VtkMeshWrapper::PointType VtkMeshWrapper::GeodesicWalk(VtkMeshWrapper::PointType pointa, int idx,
+                                                       vnl_vector_fixed<double, 3> vector) const
+{
+  //std::cerr << "------------------------------------------\n";
+  //std::cerr << "GeodesicWalk\n";
+
+  PointType snapped = this->SnapToMesh(pointa, idx);
+  vec3 point(snapped[0], snapped[1], snapped[2]);
+
+  double closest_point[3];
+
+  int faceIndex = GetTriangleForPoint(point.data(), idx, closest_point);
+
+  vec3 currentBary = this->ComputeBarycentricCoordinates(point, faceIndex);
+  //std::cerr << "Starting Bary: " << PrintValue<Eigen::Vector3d>(currentBary) << "\n";
+
+
+  Eigen::Vector3d vectorEigen = convert<vnl_vector_fixed<double, DIMENSION>, Eigen::Vector3d>(
+    vector);
+  Eigen::Vector3d projectedVector = this->ProjectVectorToFace(GetFaceNormal(faceIndex),
+                                                              vectorEigen);
+
+  Eigen::Vector3d snappedPoint = convert<PointType, Eigen::Vector3d>(snapped);
+  int ending_face = -1;
+
+  Eigen::Vector3d newPoint = GeodesicWalkOnFace(snappedPoint, projectedVector, faceIndex,
+                                                ending_face);
+
+  PointType newPointpt;
+  newPointpt[0] = newPoint[0];
+  newPointpt[1] = newPoint[1];
+  newPointpt[2] = newPoint[2];
```

Comment:
can probably use that convert<..> method. and in a bunch of other places.

Reply:
Yes, indeed, will change.